acc/components/guide.py

```diff
@@ -0,0 +1,223 @@
+
+import math
+from mcni.AbstractComponent import AbstractComponent
+from mcni.neutron_storage import neutrons_as_npyarr, ndblsperneutron
+# import mcvine
+# import mcvine.components as mc
+import numpy
+from mcvine.acc.geometry.plane import Plane
+
+category = 'optics'
+
+
+class Guide(AbstractComponent):
+
+    @classmethod
+    def __new_plane(cls, point_p, point_q, point_r):
+        # Represent everything as numpy arrays in preparation for
+        # vectorized operations across multi-dimensional arrays.
+        return Plane.construct(numpy.array(point_p, dtype=float),
+                               numpy.array(point_q, dtype=float),
+                               numpy.array(point_r, dtype=float))
+
+    def __init__(
+            self, name,
+            w1, h1, w2, h2, l,
+            # To do: take any notice of these arguments.
+            R0=0.99, Qc=0.0219, alpha=6.07, m=2, W=0.003):
+
+        """Initialize this Guide component.
+        The guide is centered on the z-axis
+        with the entrance at z=0.
+
+        Parameters:
+        name (str): the name of this component
+        w1 (m): width at the guide entry
+        h1 (m): height at the guide entry
+        w2 (m): width at the guide exit
+        h2 (m): height at the guide exit
+        l (m): length of guide
+        R0: low-angle reflectivity
+        Qc: critical scattering vector
+        alpha: slope of reflectivity
+        m: m-value of material (0 is complete absorption)
+        W: width of supermirror cutoff
+        """
+
+        self.name = name
+        self.sides = [
+            Guide.__new_plane([+w2/2, -h2/2, l],
+                              [-w2/2, +h2/2, l],
+                              [-w2/2, -h2/2, l]),
+            Guide.__new_plane([+w1/2, +h1/2, 0],
+                              [+w1/2, -h1/2, 0],
+                              [+w2/2, +h2/2, l]),
+            Guide.__new_plane([-w1/2, +h1/2, 0],
+                              [-w1/2, -h1/2, 0],
+                              [-w2/2, +h2/2, l]),
+            Guide.__new_plane([+w1/2, +h1/2, 0],
+                              [-w1/2, +h1/2, 0],
+                              [+w2/2, +h2/2, l]),
+            Guide.__new_plane([+w1/2, -h1/2, 0],
+                              [-w1/2, -h1/2, 0],
+                              [+w2/2, -h2/2, l])
+            ]
+        self.entrance = Guide.__new_plane([+w1/2, +h1/2, 0],
+                                          [+w1/2, -h1/2, 0],
+                                          [-w1/2, -h1/2, 0])
+        self.w1 = w1
+        self.h1 = h1
+        self.R0 = R0
+        self.Qc = Qc
+        self.alpha = alpha
+        self.m = m
+        self.W = W
+        self.m_neutron = 1.67492e-27  # mass of neutron in kg (from mcstas manual)
+        self.hbar = 1.05459e-34       # planck constant in Js (from mcstas manual)
+        self.m_over_h = 1e-10 * self.m_neutron / self.hbar  # includes A^-1 to m conversion
```

Comment:
Compare with `V2K` in B6 of the McStas manual.
```suggestion
        self.m_over_h = 1e10 * self.m_neutron / self.hbar  # includes A^-1 to m conversion
```
However, I am confused. If I run through the numbers with, say,
```python
from mcvine.acc.components.guide import Guide
import numpy as np
m_neutron = 1.67492e-27
hbar = 1.05459e-34
m_over_h = 1e10 * m_neutron / hbar
R0 = 0.99
Qc = 0.0219  # Å-1
alpha = 6.07  # Å
m = 2
W = 0.003  # Å-1
guide = Guide('test guide', 3, 3, 2, 2, 16)
v_i = np.array([300, 500, 700], dtype=float)  # m/s
v_f = guide.sides[1].reflect(v_i)  # m/s
(k_i, k_f) = map(lambda v: m_over_h * v, (v_i, v_f))  # Å-1
Q = np.linalg.norm(k_i - k_f)  # Å-1
np.tanh((Q - m * Qc) / W)
```
then Q is 1.02e20 so, by section 5.1.1 of the manual, the output of that last is 1.0 so the reflectivity is 0.0.

However, if we keep with @ckendrick's original here, Q comes out at a saner-looking 1.02-ish. Does that look more plausible, @yxqd, maybe there's a misprint in the McStas manual?

Unfortunately, even with the code's original 1e-10, the `np.tanh` *still* comes out as 1.0 for me. Maybe I picked a bad example to start with? Are my neutrons too fast or the reflection angle wrong?

Reply:
I should add, I do manage to get sub-1.0 value with,
```python
m_over_h = 1e-10 * m_neutron / hbar
v_i = np.array([5, 7, 300], dtype=float)  # m/s
```

Reply:
m_over_h don't need a 1e10 I think. Should be 

`m_over_h = m_neutron / hbar`

Q should be in the order of 1/angstrom, or 1e10 /m. 

For converting velocity (in m/s) to Q (in 1/angstrom), you could do

```
from mcni.utils import conversion
q = conversion.v2k(v)
```

Reply:
If you want to include the conversion from 1/m to 1/angstrom in `m_over_h`, it should be `m_over_h = 1e-10 * m_neutron / hbar`

Reply:
The `1.0e-10` conversion factor was added to convert from angstroms to meters (eq. 5.2 in the manual mentions Q is the length of the scattering vector in Angstroms^-1). Looking over your example, this makes me think that we are not also including this conversion in `Qc`?
I could also be totally wrong about even needing this conversion in the first place.

Reply:
neutron velocity should be quite parallel to the mirror in order for it to be reflected. so v should be something like [1, 0, 500], so that the majority of the component is along z direction (z is assumed to the beam direction)

Reply:
how about a quick meeting @ckendrick @mtbc

Reply:
Yes, in the above scratch code, the `m_over_h` is misnamed to try to match what @ckendrick's doing, it in fact corresponds to the `V2K` of B6 of the manual in order to effect the conversion suggested in equation 5.2.

Reply:
Some quick notes:

1. I was assuming that neutron's velocity is in m/s because the guide's dimensions are in m/s.
1. B6 of the McStas manual defines `vtok` as including a factor of 1e10 to translate from m/s to Å-1, probably because of the following,
1. 5.11 and 5.2 suggest that Q, Qc, W are in Å-1.
1. `mcni.utils.conversion.v2k` has the mass/planck factor but *not* the 1e10 scale factor.

If we take the constants and units mentioned in the manual seriously then my guess is that either:
* the velocity cannot be in m/s to start with, or,
* only a *really* small momentum change for the neutron admits of any reflection.

Reply:
Summary turns out to be that the McStas manual may well be imperfect!